File: javae/src/IceInternal/Outgoing.java
// **********************************************************************
//
// Copyright (c) 2003-2007 ZeroC, Inc. All rights reserved.
//
// This copy of Ice-E is licensed to you under the terms described in the
// ICEE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceInternal;

public final class Outgoing
{
    public
    Outgoing(Ice.Connection connection, Reference ref, String operation, Ice.OperationMode mode,
	     java.util.Hashtable context)
    {
        _connection = connection;
        _reference = ref;
        _state = StateUnsent;
        _stream = new BasicStream(ref.getInstance());

        writeHeader(operation, mode, context);
    }

    //
    // This function allows this object to be reused, rather than reallocated.
    //
    public void
    reset(Reference ref, String operation, Ice.OperationMode mode, java.util.Hashtable context)
    {
	_reference = ref;
        _state = StateUnsent;
        _exception = null;

	if(IceUtil.Debug.ASSERT)
	{
	    IceUtil.Debug.Assert(_stream != null);
	}
        _stream.reset();

        writeHeader(operation, mode, context);
    }

    // Returns true if ok, false if user exception.
    public boolean
    invoke()
        throws LocalExceptionWrapper
    {
	if(IceUtil.Debug.ASSERT)
	{
	    IceUtil.Debug.Assert(_state == StateUnsent);
	}

        _stream.endWriteEncaps();

	_state = StateInProgress;
        switch(_reference.getMode())
        {
            case Reference.ModeTwoway:
            {
		//
		// We let all exceptions raised by sending directly
		// propagate to the caller, because they can be
		// retried without violating "at-most-once". In case
		// of such exceptions, the connection object does not
		// call back on this object, so we don't need to lock
		// the mutex, keep track of state, or save exceptions.
		//
		_connection.sendRequest(_stream, this);

                if(_exception != null)
                {
		    //
		    // TODO- what we want to do is fill in the
		    // exception's stack trace, but there doesn't seem
		    // to be a way to do this yet in CLDC.
		    //
		    //_exception.fillInStackTrace();
		    
                    //      
                    // A CloseConnectionException indicates graceful
                    // server shutdown, and is therefore always repeatable
                    // without violating "at-most-once". That's because by
                    // sending a close connection message, the server
                    // guarantees that all outstanding requests can safely
                    // be repeated.
                    //
		    // An ObjectNotExistException can always be retried as
		    // well without violating "at-most-once".
		    //
                    if(_exception instanceof Ice.CloseConnectionException || 
		       _exception instanceof Ice.ObjectNotExistException)
                    {
                        throw _exception;
                    }

                    //
                    // Throw the exception wrapped in a LocalExceptionWrapper, to
                    // indicate that the request cannot be resent without
                    // potentially violating the "at-most-once" principle.
                    //
                    throw new LocalExceptionWrapper(_exception, false);
                }

                if(_state == StateUserException)
                {
                    return false;
                }
		
		if(IceUtil.Debug.ASSERT)
		{
		    IceUtil.Debug.Assert(_state == StateOK);
		}
                break;
            }

            case Reference.ModeOneway:
            {
		//
		// For oneway requests, the connection object
                // never calls back on this object. Therefore we don't
                // need to lock the mutex or save exceptions. We simply
                // let all exceptions from sending propagate to the
                // caller, because such exceptions can be retried without
                // violating "at-most-once".
		//
		_connection.sendRequest(_stream, null);
                break;
            }

            case Reference.ModeBatchOneway:
            {
		//
		// For batch oneways the same rules as for
		// regular oneways (see comment above)
		// apply.
		//
                _connection.finishBatchRequest(_stream);
                break;
            }

	    case Reference.ModeDatagram:
	    case Reference.ModeBatchDatagram:
	    {
		if(IceUtil.Debug.ASSERT)
		{
	    	    IceUtil.Debug.Assert(false);
		}
	        return false;
	    }
        }

        return true;
    }

    public void
    abort(Ice.LocalException ex)
        throws LocalExceptionWrapper
    {
	if(IceUtil.Debug.ASSERT)
	{
	    IceUtil.Debug.Assert(_state == StateUnsent);
	}

	//
	// If we didn't finish a batch oneway request, we
	// must notify the connection about that we give up ownership
	// of the batch stream.
	//
	int mode = _reference.getMode();
	if(mode == Reference.ModeBatchOneway)
	{
	    _connection.abortBatchRequest();

	    //
	    // If we abort a batch requests, we cannot retry, because
	    // not only the batch request that caused the problem will
	    // be aborted, but all other requests in the batch as
	    // well.
	    //
            throw new LocalExceptionWrapper(ex, false);
	}

	throw ex;
    }

    public void
    finished(BasicStream is)
    {
	if(IceUtil.Debug.ASSERT)
	{
	    IceUtil.Debug.Assert(_reference.getMode() == Reference.ModeTwoway); // Can only be called for twoways.
	
	    IceUtil.Debug.Assert(_state <= StateInProgress);
	}
	
	//
	// Only swap the stream if the given stream is not this Outgoing object stream!
	//
	if(is != _stream)
	{
	    _stream.swap(is);
	}

	byte replyStatus = _stream.readByte();
	
	switch(replyStatus)
	{
	    case ReplyStatus.replyOK:
	    {
		//
		// Input and output parameters are always sent in an
		// encapsulation, which makes it possible to forward
		// oneway requests as blobs.
		//
		_stream.startReadEncaps();
		_state = StateOK; // The state must be set last, in case there is an exception.
		break;
	    }
	    
	    case ReplyStatus.replyUserException:
	    {
		//
		// Input and output parameters are always sent in an
		// encapsulation, which makes it possible to forward
		// oneway requests as blobs.
		//
		_stream.startReadEncaps();
		_state = StateUserException; // The state must be set last, in case there is an exception.
		break;
	    }
	    
	    case ReplyStatus.replyObjectNotExist:
	    case ReplyStatus.replyFacetNotExist:
	    case ReplyStatus.replyOperationNotExist:
	    {
		Ice.RequestFailedException ex = null;
		switch(replyStatus)
		{
		    case ReplyStatus.replyObjectNotExist:
		    {
			ex = new Ice.ObjectNotExistException();
			break;
		    }
		    
		    case ReplyStatus.replyFacetNotExist:
		    {
			ex = new Ice.FacetNotExistException();
			break;
		    }
		    
		    case ReplyStatus.replyOperationNotExist:
		    {
			ex = new Ice.OperationNotExistException();
			break;
		    }
		    
		    default:
		    {
			if(IceUtil.Debug.ASSERT)
			{
			    IceUtil.Debug.Assert(false);
			}
			break;
		    }
		}
		
		ex.id = new Ice.Identity();
		ex.id.__read(_stream);

                //
                // For compatibility with the old FacetPath.
                //
                String[] facetPath = _stream.readStringSeq();
                if(facetPath.length > 0)
                {
		    if(facetPath.length > 1)
		    {
		        throw new Ice.MarshalException();
		    }
                    ex.facet = facetPath[0];
                }
		else
		{
		    ex.facet = "";
		}

		ex.operation = _stream.readString();
		_exception = ex;

		_state = StateLocalException; // The state must be set last, in case there is an exception.
		break;
	    }
	    
	    case ReplyStatus.replyUnknownException:
	    case ReplyStatus.replyUnknownLocalException:
	    case ReplyStatus.replyUnknownUserException:
	    {
		Ice.UnknownException ex = null;
		switch(replyStatus)
		{
		    case ReplyStatus.replyUnknownException:
		    {
			ex = new Ice.UnknownException();
			break;
		    }
		    
		    case ReplyStatus.replyUnknownLocalException:
		    {
			ex = new Ice.UnknownLocalException();
			break;
		    }
		    
		    case ReplyStatus.replyUnknownUserException: 
		    {
			ex = new Ice.UnknownUserException();
			break;
		    }
		    
		    default:
		    {
			if(IceUtil.Debug.ASSERT)
			{
			    IceUtil.Debug.Assert(false);
			}
			break;
		    }
		}
		
		ex.unknown = _stream.readString();
		_exception = ex;

		_state = StateLocalException; // The state must be set last, in case there is an exception.
		break;
	    }
	    
	    default:
	    {
		_exception = new Ice.ProtocolException("unknown reply status");
		_state = StateLocalException;
		break;
	    }
	}
    }

    public void
    finished(Ice.LocalException ex)
    {
	if(IceUtil.Debug.ASSERT)
	{
	    IceUtil.Debug.Assert(_reference.getMode() == Reference.ModeTwoway); // Can only be called for twoways.
	    IceUtil.Debug.Assert(_state <= StateInProgress);
	}

	_state = StateLocalException;
	_exception = ex;
    }

    public BasicStream
    stream()
    {
        return _stream;
    }

    private void
    writeHeader(String operation, Ice.OperationMode mode, java.util.Hashtable context)
    {
        switch(_reference.getMode())
        {
            case Reference.ModeTwoway:
            case Reference.ModeOneway:
            {
                _stream.writeBlob(Protocol.requestHdr);
                break;
            }

            case Reference.ModeBatchOneway:
            {
                _connection.prepareBatchRequest(_stream);
                break;
            }

	    case Reference.ModeDatagram:
	    case Reference.ModeBatchDatagram:
	    {
		if(IceUtil.Debug.ASSERT)
		{
	    	    IceUtil.Debug.Assert(false);
		}
	        break;
	    }
        }

        _reference.getIdentity().__write(_stream);

        //
        // For compatibility with the old FacetPath.
        //
	String facet = _reference.getFacet();
        if(facet == null || facet.length() == 0)
        {
            _stream.writeStringSeq(null);
        }
        else
        {
            String[] facetPath = { facet };
            _stream.writeStringSeq(facetPath);
        }

        _stream.writeString(operation);

        _stream.writeByte((byte)mode.value());

        if(context == null)
        {
            context = _reference.getContext();
        }

        final int sz = context.size();
        _stream.writeSize(sz);
        if(sz > 0)
        {
	    java.util.Enumeration e = context.keys();
            while(e.hasMoreElements())
            {
		String key = (String)e.nextElement();
		String value = (String)context.get(key);
                _stream.writeString(key);
                _stream.writeString(value);
            }
        }

        //
        // Input and output parameters are always sent in an
        // encapsulation, which makes it possible to forward requests as
        // blobs.
        //
        _stream.startWriteEncaps();
    }

    public int
    state()
    {
        return _state;
    }

    private Ice.Connection _connection;
    private Reference _reference;
    private Ice.LocalException _exception;

    public static final int StateUnsent = 0;
    public static final int StateInProgress = 1;
    public static final int StateOK = 2;
    public static final int StateUserException = 3;
    public static final int StateLocalException = 4;
    private int _state;

    private BasicStream _stream;

    public Outgoing next;
}

File: cppe/src/IceE/Outgoing.cpp
// **********************************************************************
//
// Copyright (c) 2003-2007 ZeroC, Inc. All rights reserved.
//
// This copy of Ice-E is licensed to you under the terms described in the
// ICEE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <IceE/LocalException.h> // Need to be included before Outgoing.h because of std::auto_ptr<LocalException>
#include <IceE/Outgoing.h>
#include <IceE/ReplyStatus.h>
#include <IceE/Connection.h>
#include <IceE/Reference.h>
#include <IceE/Instance.h>
#include <IceE/Protocol.h>

using namespace std;
using namespace Ice;
using namespace IceInternal;

IceInternal::LocalExceptionWrapper::LocalExceptionWrapper(const LocalException& ex, bool r) :
    _retry(r)
{
    _ex.reset(dynamic_cast<LocalException*>(ex.ice_clone()));
}

IceInternal::LocalExceptionWrapper::LocalExceptionWrapper(const LocalExceptionWrapper& ex) :
    _retry(ex._retry)
{
    _ex.reset(dynamic_cast<LocalException*>(ex.get()->ice_clone()));
}

const LocalException*
IceInternal::LocalExceptionWrapper::get() const
{
    assert(_ex.get());
    return _ex.get();
}

bool
IceInternal::LocalExceptionWrapper::retry() const
{
    return _retry;
}

IceInternal::Outgoing::Outgoing(Connection* connection, Reference* ref, const string& operation,
				OperationMode mode, const Context* context) :
    _connection(connection),
    _reference(ref),
    _state(StateUnsent),
    _stream(ref->getInstance().get(), ref->getInstance()->messageSizeMax()
#ifdef ICEE_HAS_WSTRING
            , ref->getInstance()->initializationData().stringConverter,
            ref->getInstance()->initializationData().wstringConverter
#endif
           )
{
    switch(_reference->getMode())
    {
	case ReferenceModeTwoway:
	case ReferenceModeOneway:
	{
	    _stream.writeBlob(requestHdr, sizeof(requestHdr));
	    break;
	}

	case ReferenceModeBatchOneway:
#ifdef ICEE_HAS_BATCH
	{
	    _connection->prepareBatchRequest(&_stream);
	    break;
	}
#endif
	case ReferenceModeDatagram:
	case ReferenceModeBatchDatagram:
	{
	    assert(false);
	    break;
	}
    }

//    _reference->getIdentity().__write(&_stream);
    _stream.write(_reference->getIdentity().name); // Directly write name for performance reasons.
    _stream.write(_reference->getIdentity().category); // Directly write category for performance reasons.

    //
    // For compatibility with the old FacetPath we still write an
    // array of strings (we don't use the basic stream string array
    // method here for performance reasons.)
    //
    if(_reference->getFacet().empty())
    {
	_stream.writeSize(0);
    }
    else
    {
	_stream.writeSize(1);
	_stream.write(_reference->getFacet());
    }

    _stream.write(operation, false);

    _stream.write(static_cast<Byte>(mode));

    if(context == 0)
    {
        context = _reference->getContext();
    }

    _stream.writeSize(Int(context->size()));
    Context::const_iterator p;
    for(p = context->begin(); p != context->end(); ++p)
    {
	_stream.write(p->first);
	_stream.write(p->second);
    }
    
    //
    // Input and output parameters are always sent in an
    // encapsulation, which makes it possible to forward requests as
    // blobs.
    //
    _stream.startWriteEncaps();
}

bool
IceInternal::Outgoing::invoke()
{
    assert(_state == StateUnsent);
    _state = StateInProgress;

    _stream.endWriteEncaps();
    
    switch(_reference->getMode())
    {
	case ReferenceModeTwoway:
	{
	    //
	    // We let all exceptions raised by sending directly
	    // propagate to the caller, because they can be retried
	    // without violating "at-most-once". In case of such
	    // exceptions, the connection object does not call back on
	    // this object, so we don't need to lock the mutex, keep
	    // track of state, or save exceptions.
	    //
	    _connection->sendRequest(&_stream, this);

	    if(_exception.get())
	    {
		//
		// A CloseConnectionException indicates graceful
		// server shutdown, and is therefore always repeatable
		// without violating "at-most-once". That's because by
		// sending a close connection message, the server
		// guarantees that all outstanding requests can safely
		// be repeated.
		//
		// An ObjectNotExistException can always be retried as
		// well without violating "at-most-once".
		//
		if(dynamic_cast<CloseConnectionException*>(_exception.get()) ||
		   dynamic_cast<ObjectNotExistException*>(_exception.get()))
		{
		    _exception->ice_throw();
		}
		
		//
		// Throw the exception wrapped in a LocalExceptionWrapper, to
		// indicate that the request cannot be resent without
		// potentially violating the "at-most-once" principle.
		//
		throw LocalExceptionWrapper(*_exception.get(), false);
	    }
	    
	    if(_state == StateUserException)
	    {
		return false;
	    }

	    assert(_state == StateOK);
	    break;
	}
	
	case ReferenceModeOneway:
	{
	    //
	    // For oneway requests, the connection object
	    // never calls back on this object. Therefore we don't
	    // need to lock the mutex or save exceptions. We simply
	    // let all exceptions from sending propagate to the
	    // caller, because such exceptions can be retried without
	    // violating "at-most-once".
	    //
	    _connection->sendRequest(&_stream, 0);
	    break;
	}

	case ReferenceModeBatchOneway:
#ifdef ICEE_HAS_BATCH
	{
	    //
	    // For batch oneways, the same rules as for
	    // regular oneways (see comment above)
	    // apply.
	    //
	    _connection->finishBatchRequest(&_stream);
	    break;
	}
#endif
	case ReferenceModeDatagram:
	case ReferenceModeBatchDatagram:
	{
	    assert(false);
	    return false;
	}
    }

    return true;
}

void
IceInternal::Outgoing::abort(const LocalException& ex)
{
    assert(_state == StateUnsent);
    
    //
    // If we didn't finish a batch oneway request, we must
    // notify the connection about that we give up ownership of the
    // batch stream.
    //
#ifdef ICEE_HAS_BATCH
    if(_reference->getMode() == ReferenceModeBatchOneway)
    {
	_connection->abortBatchRequest();
	
	//
	// If we abort a batch requests, we cannot retry, because not
	// only the batch request that caused the problem will be
	// aborted, but all other requests in the batch as well.
	//
	throw LocalExceptionWrapper(ex, false);
    }
#endif
    
    ex.ice_throw();
}

void
IceInternal::Outgoing::finished(BasicStream& is)
{
    assert(_reference->getMode() == ReferenceModeTwoway); // Can only be called for twoways.
    assert(_state <= StateInProgress);

    //
    // Only swap the stream if the given stream is not this Outgoing object stream!
    //
    if(&is != &_stream)
    {
	_stream.swap(is);
    }

    Byte replyStatus;
    _stream.read(replyStatus);
    
    switch(replyStatus)
    {
	case replyOK:
	{
	    //
	    // Input and output parameters are always sent in an
	    // encapsulation, which makes it possible to forward
	    // oneway requests as blobs.
	    //
	    _stream.startReadEncaps();
	    _state = StateOK; // The state must be set last, in case there is an exception.
	    break;
	}
	
	case replyUserException:
	{
	    //
	    // Input and output parameters are always sent in an
	    // encapsulation, which makes it possible to forward
	    // oneway requests as blobs.
	    //
	    _stream.startReadEncaps();
	    _state = StateUserException; // The state must be set last, in case there is an exception.
	    break;
	}
	
	case replyObjectNotExist:
	case replyFacetNotExist:
	case replyOperationNotExist:
	{
	    //
	    // Don't read the exception members directly into the
	    // exception. Otherwise if reading fails and raises an
	    // exception, you will have a memory leak.
	    //
	    Identity ident;
	    ident.__read(&_stream);

	    //
	    // For compatibility with the old FacetPath.
	    //
	    vector<string> facetPath;
	    _stream.read(facetPath);
	    string facet;
	    if(!facetPath.empty())
	    {
		if(facetPath.size() > 1)
		{
		    throw MarshalException(__FILE__, __LINE__);
		}
		facet.swap(facetPath[0]);
	    }

	    string operation;
	    _stream.read(operation, false);
	    
	    RequestFailedException* ex;
	    switch(replyStatus)
	    {
		case replyObjectNotExist:
		{
		    ex = new ObjectNotExistException(__FILE__, __LINE__);
		    break;
		}
		
		case replyFacetNotExist:
		{
		    ex = new FacetNotExistException(__FILE__, __LINE__);
		    break;
		}
		
		case replyOperationNotExist:
		{
		    ex = new OperationNotExistException(__FILE__, __LINE__);
		    break;
		}
		
		default:
		{
		    ex = 0; // To keep the compiler from complaining.
		    assert(false);
		    break;
		}
	    }
	    
	    ex->id = ident;
	    ex->facet = facet;
	    ex->operation = operation;
	    _exception.reset(ex);

	    _state = StateLocalException; // The state must be set last, in case there is an exception.
	    break;
	}
	
	case replyUnknownException:
	case replyUnknownLocalException:
	case replyUnknownUserException:
	{
	    //
	    // Don't read the exception members directly into the
	    // exception. Otherwise if reading fails and raises an
	    // exception, you will have a memory leak.
	    //
	    string unknown;
	    _stream.read(unknown, false);
	    
	    UnknownException* ex;
	    switch(replyStatus)
	    {
		case replyUnknownException:
		{
		    ex = new UnknownException(__FILE__, __LINE__);
		    break;
		}
		
		case replyUnknownLocalException:
		{
		    ex = new UnknownLocalException(__FILE__, __LINE__);
		    break;
		}
		
		case replyUnknownUserException:
		{
		    ex = new UnknownUserException(__FILE__, __LINE__);
		    break;
		}
		
		default:
		{
		    ex = 0; // To keep the compiler from complaining.
		    assert(false);
		    break;
		}
	    }
	    
	    ex->unknown = unknown;
	    _exception.reset(ex);

	    _state = StateLocalException; // The state must be set last, in case there is an exception.
	    break;
	}
	
	default:
	{
	    //_exception.reset(new UnknownReplyStatusException(__FILE__, __LINE__));
	    _exception.reset(new ProtocolException(__FILE__, __LINE__, "unknown reply status"));
	    _state = StateLocalException;
	    break;
	}
    }
}

void
IceInternal::Outgoing::finished(const LocalException& ex)
{
    assert(_reference->getMode() == ReferenceModeTwoway); // Can only be called for twoways.
    assert(_state <= StateInProgress);
    
    _state = StateLocalException;
    _exception.reset(dynamic_cast<LocalException*>(ex.ice_clone()));
}
